Vagrantfile

```diff
@@ -18,11 +18,96 @@ end
 ENV['ANSIBLE_CONFIG'] = ansible_config_file
 
 NMONS        = ENV["VMS"] || settings['vms']
-SUBNET       = settings['subnet']
 BOX          = settings['vagrant_box']
 BOX_VERSION  = settings['box_version']
 memory       = settings['memory']
 
+SUBNET_PREFIX          = settings['subnet_prefix']
+SUBNET_ASSIGNMENT_FILE = "subnet_assignment.state"
+SUBNET_STATE_FILE      = "/tmp/volplugin_vagrant_subnets.state"
+SUBNET_STATE_LOCKFILE  = SUBNET_STATE_FILE + ".lock"
+
+class SubnetStateFileError < Exception; end
+class NoAvailableSubnetsError < Exception; end
+
+# This method allocates a random subnet for Vagrant to use by combining the prefix
+# specified in vagrant_variables.yml with a random third octet guaranteed to be
+# unused by other environments.  If a previous subnet was allocated, it will be used.
+#
+# The subnet reservation process is locked, so multiple environments can be brought
+# up in parallel with no issues.
+def random_subnet
+  if File.exists?(SUBNET_ASSIGNMENT_FILE)
+
+    # only use the existing assignment if the state file still exists
+    # i.e., the machine hasn't been rebooted
+    if File.exists?(SUBNET_STATE_FILE)
+      octet = File.read(SUBNET_ASSIGNMENT_FILE).strip
+
+      subnet_mask = "#{SUBNET_PREFIX}#{octet}"
+      puts "Returning: #{subnet_mask}"
+      return subnet_mask
+    else
+      msg = [
+        "Subnet state file is missing.",
+        "Are you trying to re-use an existing environment after a reboot?",
+        "To continue, you should delete #{SUBNET_ASSIGNMENT_FILE} in the current directory",
+        "or rebuild the environment from scratch."
+      ]
+      raise SubnetStateFileError.new(msg.join(" "))
+    end
+  end
+
+  # no existing subnet assignment, so reserve a new one
+
+  FileUtils.touch(SUBNET_STATE_FILE)
+  FileUtils.touch(SUBNET_STATE_LOCKFILE)
+
+  f = File.open(SUBNET_STATE_LOCKFILE)
```

Comment:
you're gonna hate me, but you can chuck probably 50% of this code if you just use a directory as your database with the randomized octet as filenames in it. Locking is for free that way.